turbopack/crates/turbo-tasks-backend/src/backend/operation/connect_children.rs

```diff
@@ -35,39 +37,117 @@ pub fn connect_children(
+
+        let mut queue = AggregationUpdateQueue::new();
+
+        if let Some(upper_ids) = upper_ids {
+            // We need to add new followers when there are upper ids as the parent is a leaf node
+            // and new children are new followers.
+            let add_followers = !upper_ids.is_empty();
+
+            // And we need decrease the active count when parent doesn't have an active count as the
+            // active count was temporarily increased during connect_child. We need to
+            // increase the active count when the parent has active count, because it's
+            // added as follower.
+            let decrease_active_count = should_track_activeness && !has_active_count;
+
+            // We special case the situation when we need to do both operations to avoid
+            // cloning the new follower ids unnecessarily.
+            if decrease_active_count && add_followers {
+                queue.push(
+                    InnerOfUppersHasNewFollowersJob {
+                        upper_ids,
+                        new_follower_ids: new_follower_ids.clone(),
+                    }
+                    .into(),
+                );
+                queue.push(AggregationUpdateJob::DecreaseActiveCounts {
+                    task_ids: new_follower_ids,
+                })
+            } else if decrease_active_count {
+                queue.push(AggregationUpdateJob::DecreaseActiveCounts {
+                    task_ids: new_follower_ids,
+                })
+            } else if add_followers {
+                queue.push(
+                    InnerOfUppersHasNewFollowersJob {
+                        upper_ids,
+                        new_follower_ids,
+                    }
+                    .into(),
+                );
+            }
+        } else if should_track_activeness {
+            // Parent is an aggregating node. We run the normal code to connect the children.
+            queue.push(AggregationUpdateJob::InnerOfUpperHasNewFollowers {
+                upper_id: parent_task_id,
+                new_follower_ids: new_follower_ids.clone(),
+            });
+            // We need to decrease the active count because we temporarily increased it during
+            // connect_child.
             queue.push(AggregationUpdateJob::DecreaseActiveCounts {
                 task_ids: new_follower_ids,
             });
+        } else {
+            // Parent is an aggregating node. We run the normal code to connect the children.
+            queue.push(AggregationUpdateJob::InnerOfUpperHasNewFollowers {
+                upper_id: parent_task_id,
+                new_follower_ids,
+            });
+        }
+
+        {
+            #[cfg(feature = "trace_task_completion")]
+            let _span = tracing::trace_span!("connect new children").entered();
+            queue.execute(ctx);
         }
     }
+
+    const MIN_CHILDREN_FOR_PARALLEL: usize = 10000;
```

Comment:
nit: a brief justification would be nice.

I know we basically just care about tasks like the module graph which has like 3x number of modules tasks and a threshold like this allows us to target those usecases. So just say that?

Or is this more based on the overhead of scoping work here?

Reply:
i would also accept a justification in the form of

// 10000 ~= distance from the earth to the moon in kilometers / hitpoints of pikachu

Reply:
It's about the overhead of parallelization.
I think we should do a benchmark to measure the overhead and make up a number based on that.